### PR TITLE
Caiman_crossRegister

### PR DESCRIPTION
**Pull request by kapilseh**

### CODE REVIEW

File: CaImAn/caiman_main.py [hunk 1]
```diff
@@ -46,6 +47,16 @@ def main(caiman_parameters):
...
     Yr, dims, T, cnm, images = cnmf(n_processes, dview, fname_new, caiman_parameters)
 
+    # CaimAn Cross register
+    A = cnm.estimates.A[:,cnm.estimates.idx_components]
+    C = cnm.estimates.C[cnm.estimates.idx_components,:]
+    AC = A.dot(C)
+    shape = (dims[0],dims[1],T)
+    AC = AC.reshape(shape, order='F').transpose((-1, 0, 1))
+    A = A.toarray()
+    A = A.reshape(shape[0],shape[1],A.shape[1], order='F').transpose((-1, 0, 1))
```
Comment (4D42) [inline CaImAn/caiman_main.py:57]: it is strange that you need to do the reshaping

File: CaImAn/caiman_main.py [hunk 2]
```diff
@@ -46,6 +47,16 @@ def main(caiman_parameters):
...
+    A = cross_register(A, AC, caiman_parameters)
```
Comment (4D42) [inline CaImAn/caiman_main.py:58]: in line 78 it is saving `cm.estimates.A[...]` so I'm not sure `A = cross_register(A, AC, caiman_parameters)` will work

File: CaImAn/caiman_main.py [hunk 3]
```diff
@@ -46,6 +47,15 @@ def main(caiman_parameters):
 
     dims, T, cnm, images = cnmf(n_processes, dview, fname_new, caiman_parameters)
 
+    # CaimAn Cross register
+    A = cnm.estimates.A[:,cnm.estimates.idx_components]
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:51]: lines 51-56 have to be part of the `cross_register()` function

File: CaImAn/caiman_main.py [hunk 4]
```diff
@@ -273,3 +284,75 @@ def save_caiman_to_doric(
...
 
+def cross_register(A, AC, caiman_parameters):
+
+    if not caiman_parameters.params_cross_reg:
+        return
+
+    print(cm_defs.Messages.CROSS_REGISTRATING , flush=True)
+
+    # Load AC componenets from the reference file
+    ref_filepath = caiman_parameters.params_cross_reg["fname"]
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:295]: are `caiman_parameters` from the current file or the reference file?

File: CaImAn/caiman_main.py [hunk 5]
```diff
@@ -273,3 +284,75 @@ def save_caiman_to_doric(
...
+    ref_images   = caiman_parameters.params_cross_reg["h5path_images"]
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:296]: it's a path, I guess, not the images itself, it should be clear from the name

File: CaImAn/caiman_main.py [hunk 6]
```diff
@@ -273,3 +284,75 @@ def save_caiman_to_doric(
...
+    file_ = h5py.File(ref_filepath, 'r')
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:297]: Maybe it's better to access the file `with `

File: CaImAn/caiman_main.py [hunk 7]
```diff
@@ -273,3 +284,75 @@ def save_caiman_to_doric(
...
+    ref_images = utils.clean_path(ref_images)
+
+    if defs.DoricFile.Dataset.IMAGE_STACK in file_[ref_images]:
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:300]: Define `dataname`, and then read the data

File: CaImAn/caiman_main.py [hunk 8]
```diff
@@ -273,3 +284,75 @@ def save_caiman_to_doric(
...
+    if not caiman_parameters.params_cross_reg:
+        return
+
+    print(cm_defs.Messages.CROSS_REGISTRATING , flush=True)
+
+    # Load AC componenets from the reference file
+    ref_filepath = caiman_parameters.params_cross_reg["fname"]
+    ref_images   = caiman_parameters.params_cross_reg["h5path_images"]
+    file_ = h5py.File(ref_filepath, 'r')
+    ref_images = utils.clean_path(ref_images)
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:298]: write the lines 298 and 296 in 1 line

File: CaImAn/caiman_main.py [hunk 9]
```diff
@@ -273,3 +284,75 @@ def save_caiman_to_doric(
...
+
+    if defs.DoricFile.Dataset.IMAGE_STACK in file_[ref_images]:
+        file_image_stack = file_[f"{ref_images}/{defs.DoricFile.Dataset.IMAGE_STACK}"]
+    else:
+        file_image_stack= file_[f"{ref_images}/{defs.DoricFile.Deprecated.Dataset.IMAGES_STACK}"]
+
+    AC_ref = np.array(file_image_stack)
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:305]: `np.array(file_[`path`])`

File: CaImAn/caiman_main.py [hunk 10]
```diff
@@ -273,3 +284,75 @@ def save_caiman_to_doric(
...
+        if np.isfinite(current) and np.isfinite(ref):
+            ids[i] = roi_ids_ref[int(ref)]
+            # ids[i] = ref + 1
+        elif np.isfinite(current):
+            ids[i] = ref_id_max
+            ref_id_max += 1
+    return ids
+
+
+def get_footprints(filename, rois_h5path, refShape):
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:343]: Can we move the function to the `utilities.py`, and use for both caiman and minian?

Comment (kapilseh) [reply CaImAn/caiman_main.py:350]: I guess yes. The only diff is that in minian the output is dataarray, while in caiman the output is an array and a list . 
I can move the function to utilities with return as array and list, and then in minian I can construct the datarray in the cross register function.

Comment (doriclenses) [reply CaImAn/caiman_main.py:343]: oki, maybe later

File: CaImAn/caiman_main.py [hunk 11]
```diff
@@ -266,3 +275,84 @@ def save_caiman_to_doric(
...
+
+    print(cm_defs.Messages.CROSS_REGISTRATING , flush=True)
+
+    # Load AC componenets from the reference file
+    ref_filepath = caiman_parameters.params_cross_reg["fname"]
+    ref_images_path = utils.clean_path(caiman_parameters.params_cross_reg["h5path_images"])
+
+    with h5py.File(ref_filepath, 'r') as file_:
+        if defs.DoricFile.Dataset.IMAGE_STACK in file_[ref_images_path]:
+            path = f"{ref_images_path}/{defs.DoricFile.Dataset.IMAGE_STACK}"
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:300]: use `dataname`

File: CaImAn/caiman_main.py [hunk 12]
```diff
@@ -266,3 +275,84 @@ def save_caiman_to_doric(
...
+
+        A = A.reshape(-1, A.shape[2])
+        A_ref = A_ref.reshape(-1, A_ref.shape[2])
+        spatial = [A, A_ref]
+
+        spatial_union, assignments, matchings = register_multisession(A=spatial, dims=dims, templates=templates)
+
+        # Update unit ids of the current spatial componenets A
+        ids = [0] * A.shape[1]
+        # ref_id_max = int(A_ref.shape[1]) + 1
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:328]: can be deleted?

**Comment by doriclenses** [comment]
Hey @kapilseh, you ran some tests for minian, and posted the results in the pull request. Could you run the same tests for caiman, and display them here?

File: CaImAn/caiman_main.py [hunk 13]
```diff
@@ -266,3 +275,84 @@ def save_caiman_to_doric(
...
+        ref_id_max = int(max(roi_ids_ref)) + 1
+        for i in range(spatial_union.shape[1]):
+            current = assignments[i,0]
+            ref = assignments[i,1]
+            if np.isfinite(current) and np.isfinite(ref):
+                ids[i] = roi_ids_ref[int(ref)]
+                # ids[i] = ref + 1
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:335]: can be deleted?

**Review by doriclenses — CHANGES_REQUESTED** [review]

**Comment by kapilseh** [comment]
Test run to show the registration is working fine->
image

image

**Comment by doriclenses** [comment]
Looks good! What if you use the current file from the top image to do cross registration again (current and reference are the same)?

**Comment by doriclenses** [comment]
Except the dataset names and usernames, looks good!

File: utilities.py [hunk 14]
```diff
@@ -453,7 +445,7 @@ def def_chunk_size(data_shape):
         if durantion < chunk_size:
             chunk_size = durantion
 
-        return  chunk_size
+        return  (chunk_size,)
```
[the remark this answers is not in the document]
Comment (kapilseh) [reply utilities.py:448]: still gives me a chunk error if I change it. So yes I need this,

**Comment by kapilseh** [comment]
All seems to work fine now, all the suggested changes have been incorporated.
in main line 208-209  is reshaping/transposing -> theoretically both lines 208-209 are not required because ultimately the final array shape after line 209 is the same as what line 207 gives. but if I remove these 2 lines the registration is not correct. So probably reshaping transpose is rearranging the array somehow that I don't get.
@4D42  if you can understand what's happening we can make some changes or else at least it's working well for now.

**Comment by doriclenses** [comment]
Most likely it rearranges differently rows and cols

File: CaImAn/caiman_main.py [hunk 15]
```diff
@@ -165,6 +173,75 @@ def cnmf(n_processes, dview, fname_new, caiman_params):
...
+        if np.isfinite(current) and np.isfinite(ref):
+            ids[i] = int(roi_ids_ref[int(ref)])
+        elif np.isfinite(current):
+            ids[i] = ref_id_max
+            ref_id_max += 1
+
+    return ids
+
+
+def get_footprints(file_, rois_h5path, refShape):
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:228]: giving open `file_` now

File: CaImAn/caiman_main.py [hunk 16]
```diff
@@ -165,6 +173,75 @@ def cnmf(n_processes, dview, fname_new, caiman_params):
...
+    C: np.ndarray,
+    shape,
+    caiman_parameters
+) -> List[int]:
+
+    if not caiman_parameters.params_cross_reg:
+        return
+    print(cm_defs.Messages.CROSS_REGISTRATING , flush=True)
+
+    # Load reference images (AC_ref) and ROI footprints (A_ref)
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:187]: I rearranged code, please make sure that I didn't miss anything

File: CaImAn/caiman_main.py [hunk 17]
```diff
@@ -165,6 +173,72 @@ def cnmf(n_processes, dview, fname_new, caiman_params):
...
+    shape = (shape[1], shape[2], shape[0]);
+    AC = (A.dot(C)).reshape(shape, order='F')  # reshaping from caiman shape (2D) to image shape (3D)
+    AC_max  = AC.max(axis = 2)
+    AC_ref = AC_ref.astype(float)
+    AC_ref_max = AC_ref.max(axis = 2)
+    templates = [AC_ref_max, AC_max]
+
+    # Concatenate footprints (A and A_ref)
+    A_ref = A_ref.transpose(1, 2, 0).reshape((-1, A_ref.shape[0]), order='F')  # reshape to caiman shape (2D)
+    A  = A.toarray()
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:205]: do we need the `toarray()`?

File: CaImAn/caiman_main.py [hunk 18]
```diff
@@ -165,6 +173,72 @@ def cnmf(n_processes, dview, fname_new, caiman_params):
...
+    # Load reference images (AC_ref) and ROI footprints (A_ref)
+    ref_filepath = caiman_parameters.params_cross_reg["fname"]
+    ref_rois_path  = caiman_parameters.params_cross_reg["h5path_roi"]
+    ref_images_path = utils.clean_path(caiman_parameters.params_cross_reg["h5path_images"])
+    with h5py.File(ref_filepath, 'r') as file_:
+        AC_ref = np.array(file_[f"{ref_images_path}/{caiman_parameters.dataname}"])
+        A_ref, roi_ids_ref = get_footprints(file_, ref_rois_path, AC_ref.shape)
+
+    # Concatenate max proj of images (AC and AC_ref)
+    shape = (shape[1], shape[2], shape[0]);
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:196]: do we have to do it on a separate line or can we just write it inside the reshape on the line 197?

File: CaImAn/caiman_main.py [hunk 19]
```diff
@@ -165,6 +173,72 @@ def cnmf(n_processes, dview, fname_new, caiman_params):
...
+    AC = (A.dot(C)).reshape(shape, order='F')  # reshaping from caiman shape (2D) to image shape (3D)
+    AC_max  = AC.max(axis = 2)
+    AC_ref = AC_ref.astype(float)
```
Comment (doriclenses) [inline CaImAn/caiman_main.py:199]: it's not float by default? or can we do it when loading data on the line 192?

File: CaImAn/caiman_main.py (hunk 17 above)
Comment (kapilseh) [reply CaImAn/caiman_main.py:205]: yes, because initially it is a sparse matrix

**Comment by doriclenses** [comment]
@kapilseh could you ran the tests again and show the results here?

**Comment by kapilseh** [comment]
Test results-

**Comment by doriclenses** [comment]
great! does it also work well if use the current file from the top image (as a current and reference file)?

**Comment by kapilseh** [comment]
`great! does it also work well if you use the current file from the top image (as a current and reference file)?`
Yes it works

**Review by doriclenses — APPROVED** [review]
great job!

**Review by 4D42 — APPROVED** [review]

**Pull request merged**